site-src/geps/gep-724.md

```diff
@@ -0,0 +1,614 @@
+  // +kubebuilder:default=gateway.networking.k8s.io
+  Group string `json:"group"`
+
+  // Kind is kind of the referent.
+  //
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=253
+  Kind string `json:"kind"`
+
+  // Namespace is the namespace of the referent. When unspecified (empty
+  // string), this will either be:
+  //
+  // * local namespace of the target is a namespace scoped resource
+  // * no namespace (not applicable) if the target is cluster-scoped.
+  //
+  // Support: Extended
+  //
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=253
+  // +optional
+  Namespace string `json:"namespace,omitempty"`
+
+  // Name is the name of the referent.
+  //
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=253
+  Name string `json:"name"`
+
+  // SectionName is the name of a section within the target resource. In the
+  // following resources, SectionName is interpreted as the following:
+  //
+  // * Gateway: Listener Name
+  //
+  // Implementations MAY choose to support attaching Routes to other resources.
+  // If that is the case, they MUST clearly document how SectionName is
+  // interpreted.
+  //
+  // When unspecified (empty string), this will reference the entire resource.
+  // For the purpose of status, an attachment is considered successful if at
+  // least one section in the parent resource accepts it. For example, Gateway
+  // listeners can restrict which Routes can bind to them by Route kind,
+  // namespace, or hostname. If 1 of 2 Gateway listeners accept attachment from
+  // the referencing Route, the Route MUST be considered successfully
+  // attached. If no Gateway listeners accept attachment from this Route, the
+  // Route MUST be considered detached from the Gateway.
+  //
+  // Support: Core
+  //
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=253
+  // +optional
+  SectionName string `json:"sectionName,omitempty"`
+}
+```
+
+This would also require adding a `Name` field to Gateway listeners and Route
+rules to support the SectionName concept:
+
+```go
+type Listener struct {
+    // Name is the name of the Listener. If more than one Listener is present
+    // each Listener MUST specify a name. The names of Listeners MUST be unique
+    // within a Gateway.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    // +optional
+    Name string `json:"name,omitempty"`
+    // ...
+}
+```
+
+```go
+type RouteRule struct {
+    // Name is the name of the Route rule. If more than one Route Rule is
+    // present, each Rule MUST specify a name. The names of Rules MUST be unique
+    // within a Route.
+    //
+    // Support: Core
+    //
+    // +kubebuilder:validation:MinLength=1
+    // +kubebuilder:validation:MaxLength=253
+    // +optional
+    Name string `json:"name,omitempty"`
```

Comment:
Is this assuming that we will build route inclusion using this mechanism?
If that is the only reason, please remove this field. We can add it in later once we are more sure (we don't seem to be sure right now: https://github.com/kubernetes-sigs/gateway-api/pull/725/files#r677977878).

Does ReferencePolicy have anything to do here?

Reply:
I think its still used to attach policy to a route, even if not Route<->Route. Ie apply TimeoutPolicy to a route

Reply:
Yeah this was copied verbatim from the original GEP for policy attachment where this was proposed. My rationale for adding it here was so that it could also apply to policy attachment, but I'll pull that out here because it's out of scope. It will be fairly easy to add later if necessary.

Reply:
Removed this, we can add it later if needed.